Editor/Binaries/macOS/ripgrep-14.1.1-aarch64-apple-darwin/doc/rg.1
.TH RG 1 2024-09-08 "14.1.1 (rev 4649aa9700)" "User Commands"
.
.
.SH NAME
rg \- recursively search the current directory for lines matching a pattern
.
.
.SH SYNOPSIS
.\" I considered using GNU troff's .SY and .YS "synopsis" macros here, but it
.\" looks like they aren't portable. Specifically, they don't appear to be in
.\" BSD's mdoc used on macOS.
.sp
\fBrg\fP [\fIOPTIONS\fP] \fIPATTERN\fP [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-e\fP \fIPATTERN\fP... [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-f\fP \fIPATTERNFILE\fP... [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-files\fP [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-type\-list\fP
.sp
\fIcommand\fP | \fBrg\fP [\fIOPTIONS\fP] \fIPATTERN\fP
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-help\fP
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-version\fP
.
.
.SH DESCRIPTION
ripgrep (rg) recursively searches the current directory for a regex pattern.
By default, ripgrep will respect your \fB.gitignore\fP and automatically skip
hidden files/directories and binary files.
.sp
ripgrep's default regex engine uses finite automata and guarantees linear
time searching. Because of this, features like backreferences and arbitrary
look-around are not supported. However, if ripgrep is built with PCRE2,
then the \fB\-P/\-\-pcre2\fP flag can be used to enable backreferences and
look-around.
.sp
ripgrep supports configuration files. Set \fBRIPGREP_CONFIG_PATH\fP to a
configuration file. The file can specify one shell argument per line. Lines
starting with \fB#\fP are ignored. For more details, see \fBCONFIGURATION
FILES\fP below.
.sp
ripgrep will automatically detect if stdin exists and search stdin for a regex
pattern, e.g. \fBls | rg foo\fP. In some environments, stdin may exist when
it shouldn't. To turn off stdin detection, one can explicitly specify the
directory to search, e.g. \fBrg foo ./\fP.
.sp
Like other tools such as \fBls\fP, ripgrep will alter its output depending on
whether stdout is connected to a tty. By default, when printing a tty, ripgrep
will enable colors, line numbers and a heading format that lists each matching
file path once instead of once per matching line.
.sp
Tip: to disable all smart filtering and make ripgrep behave a bit more like
classical grep, use \fBrg -uuu\fP.
.
.
.SH REGEX SYNTAX
ripgrep uses Rust's regex engine by default, which documents its syntax:
\fIhttps://docs.rs/regex/1.*/regex/#syntax\fP
.sp
ripgrep uses byte-oriented regexes, which has some additional documentation:
\fIhttps://docs.rs/regex/1.*/regex/bytes/index.html#syntax\fP
.sp
To a first approximation, ripgrep uses Perl-like regexes without look-around or
backreferences. This makes them very similar to the "extended" (ERE) regular
expressions supported by *egrep*, but with a few additional features like
Unicode character classes.
.sp
If you're using ripgrep with the \fB\-P/\-\-pcre2\fP flag, then please consult
\fIhttps://www.pcre.org\fP or the PCRE2 man pages for documentation on the
supported syntax.
.
.
.SH POSITIONAL ARGUMENTS
.TP 12
\fIPATTERN\fP
A regular expression used for searching. To match a pattern beginning with a
dash, use the \fB\-e/\-\-regexp\fP option.
.TP 12
\fIPATH\fP
A file or directory to search. Directories are searched recursively. File paths
specified explicitly on the command line override glob and ignore rules.
.
.
.SH OPTIONS
This section documents all flags that ripgrep accepts. Flags are grouped into
categories below according to their function.
.sp
Note that many options can be turned on and off. In some cases, those flags are
not listed explicitly below. For example, the \fB\-\-column\fP flag (listed
below) enables column numbers in ripgrep's output, but the \fB\-\-no\-column\fP
flag (not listed below) disables them. The reverse can also exist. For example,
the \fB\-\-no\-ignore\fP flag (listed below) disables ripgrep's \fBgitignore\fP
logic, but the \fB\-\-ignore\fP flag (not listed below) enables it. These
flags are useful for overriding a ripgrep configuration file (or alias) on the
command line. Each flag's documentation notes whether an inverted flag exists.
In all cases, the flag specified last takes precedence.
.
.SS INPUT OPTIONS
\fB\-e\fP \fIPATTERN\fP, \fB\-\-regexp\fP=\fIPATTERN\fP
.RS 4
A pattern to search for. This option can be provided multiple times, where
all patterns given are searched, in addition to any patterns provided by
\fB\-f/\-\-file\fP. Lines matching at least one of the provided patterns are printed.
This flag can also be used when searching for patterns that start with a dash.
.sp
For example, to search for the literal \fB\-foo\fP:
.sp
.EX
    rg \-e \-foo
.EE
.sp
You can also use the special \fB\-\-\fP delimiter to indicate that no more
flags will be provided. Namely, the following is equivalent to the above:
.sp
.EX
    rg \-\- \-foo
.EE
.sp
When \fB\-f/\-\-file\fP or \fB\-e/\-\-regexp\fP is used, then ripgrep treats all positional
arguments as files or directories to search.
.RE
.sp
\fB\-f\fP \fIPATTERNFILE\fP, \fB\-\-file\fP=\fIPATTERNFILE\fP
.RS 4
Search for patterns from the given file, with one pattern per line. When this
flag is used multiple times or in combination with the \fB\-e/\-\-regexp\fP flag, then
all patterns provided are searched. Empty pattern lines will match all input
lines, and the newline is not counted as part of the pattern.
.sp
A line is printed if and only if it matches at least one of the patterns.
.sp
When \fIPATTERNFILE\fP is \fB-\fP, then \fBstdin\fP will be read for the
patterns.
.sp
When \fB\-f/\-\-file\fP or \fB\-e/\-\-regexp\fP is used, then ripgrep treats all positional
arguments as files or directories to search.
.RE
.sp
\fB\-\-pre\fP=\fICOMMAND\fP
.RS 4
For each input \fIPATH\fP, this flag causes ripgrep to search the standard
output of \fICOMMAND\fP \fIPATH\fP instead of the contents of \fIPATH\fP.
This option expects the \fICOMMAND\fP program to either be a path or to be
available in your \fBPATH\fP. Either an empty string \fICOMMAND\fP or the
\fB\-\-no\-pre\fP flag will disable this behavior.
.sp
.TP 12
\fBWARNING\fP
When this flag is set, ripgrep will unconditionally spawn a process for every
file that is searched. Therefore, this can incur an unnecessarily large
performance penalty if you don't otherwise need the flexibility offered by this
flag. One possible mitigation to this is to use the \fB\-\-pre-glob\fP flag to
limit which files a preprocessor is run with.
.PP
A preprocessor is not run when ripgrep is searching stdin.
.sp
When searching over sets of files that may require one of several
preprocessors, \fICOMMAND\fP should be a wrapper program which first classifies
\fIPATH\fP based on magic numbers/content or based on the \fIPATH\fP name and
then dispatches to an appropriate preprocessor. Each \fICOMMAND\fP also has its
standard input connected to \fIPATH\fP for convenience.
.sp
For example, a shell script for \fICOMMAND\fP might look like:
.sp
.EX
    case "$1" in
    *.pdf)
        exec pdftotext "$1" -
        ;;
    *)
        case $(file "$1") in
        *Zstandard*)
            exec pzstd -cdq
            ;;
        *)
            exec cat
            ;;
        esac
        ;;
    esac
.EE
.sp
The above script uses \fBpdftotext\fP to convert a PDF file to plain text. For
all other files, the script uses the \fBfile\fP utility to sniff the type of
the file based on its contents. If it is a compressed file in the Zstandard
format, then \fBpzstd\fP is used to decompress the contents to stdout.
.sp
This overrides the \fB\-z/\-\-search-zip\fP flag.
.RE
.sp
\fB\-\-pre-glob\fP=\fIGLOB\fP
.RS 4
This flag works in conjunction with the \fB\-\-pre\fP flag. Namely, when one or
more \fB\-\-pre-glob\fP flags are given, then only files that match the given set
of globs will be handed to the command specified by the \fB\-\-pre\fP flag. Any
non-matching files will be searched without using the preprocessor command.
.sp
This flag is useful when searching many files with the \fB\-\-pre\fP flag.
Namely, it provides the ability to avoid process overhead for files that
don't need preprocessing. For example, given the following shell script,
\fIpre-pdftotext\fP:
.sp
.EX
    #!/bin/sh
    pdftotext "$1" -
.EE
.sp
then it is possible to use \fB\-\-pre\fP \fIpre-pdftotext\fP \fB--pre-glob
'\fP\fI*.pdf\fP\fB'\fP to make it so ripgrep only executes the
\fIpre-pdftotext\fP command on files with a \fI.pdf\fP extension.
.sp
Multiple \fB\-\-pre-glob\fP flags may be used. Globbing rules match
\fBgitignore\fP globs. Precede a glob with a \fB!\fP to exclude it.
.sp
This flag has no effect if the \fB\-\-pre\fP flag is not used.
.RE
.sp
\fB\-z\fP, \fB\-\-search-zip\fP
.RS 4
This flag instructs ripgrep to search in compressed files. Currently gzip,
bzip2, xz, LZ4, LZMA, Brotli and Zstd files are supported. This option expects
the decompression binaries (such as \fBgzip\fP) to be available in your
\fBPATH\fP. If the required binaries are not found, then ripgrep will not
emit an error messages by default. Use the \fB\-\-debug\fP flag to see more
information.
.sp
Note that this flag does not make ripgrep search archive formats as directory
trees. It only makes ripgrep detect compressed files and then decompress them
before searching their contents as it would any other file.
.sp
This overrides the \fB\-\-pre\fP flag.
.sp
This flag can be disabled with \fB\-\-no-search-zip\fP.
.RE

.
.SS SEARCH OPTIONS
\fB\-s\fP, \fB\-\-case-sensitive\fP
.RS 4
Execute the search case sensitively. This is the default mode.
.sp
This is a global option that applies to all patterns given to ripgrep.
Individual patterns can still be matched case insensitively by using inline
regex flags. For example, \fB(?i)abc\fP will match \fBabc\fP case insensitively
even when this flag is used.
.sp
This flag overrides the \fB\-i/\-\-ignore-case\fP and \fB\-S/\-\-smart-case\fP flags.
.RE
.sp
\fB\-\-crlf\fP
.RS 4
When enabled, ripgrep will treat CRLF (\fB\\r\\n\fP) as a line terminator
instead of just \fB\\n\fP.
.sp
Principally, this permits the line anchor assertions \fB^\fP and \fB$\fP in
regex patterns to treat CRLF, CR or LF as line terminators instead of just LF.
Note that they will never match between a CR and a LF. CRLF is treated as one
single line terminator.
.sp
When using the default regex engine, CRLF support can also be enabled inside
the pattern with the \fBR\fP flag. For example, \fB(?R:$)\fP will match just
before either CR or LF, but never between CR and LF.
.sp
This flag overrides \fB\-\-null-data\fP.
.sp
This flag can be disabled with \fB\-\-no-crlf\fP.
.RE
.sp
\fB\-\-dfa-size-limit\fP=\fINUM+SUFFIX?\fP
.RS 4
The upper size limit of the regex DFA. The default limit is something generous
for any single pattern or for many smallish patterns. This should only be
changed on very large regex inputs where the (slower) fallback regex engine may
otherwise be used if the limit is reached.
.sp
The input format accepts suffixes of \fBK\fP, \fBM\fP or \fBG\fP which
correspond to kilobytes, megabytes and gigabytes, respectively. If no suffix is
provided the input is treated as bytes.
.RE
.sp
\fB\-E\fP \fIENCODING\fP, \fB\-\-encoding\fP=\fIENCODING\fP
.RS 4
Specify the text encoding that ripgrep will use on all files searched. The
default value is \fBauto\fP, which will cause ripgrep to do a best effort
automatic detection of encoding on a per-file basis. Automatic detection in
this case only applies to files that begin with a UTF-8 or UTF-16 byte-order
mark (BOM). No other automatic detection is performed. One can also specify
\fBnone\fP which will then completely disable BOM sniffing and always result
in searching the raw bytes, including a BOM if it's present, regardless of its
encoding.
.sp
Other supported values can be found in the list of labels here:
\fIhttps://encoding.spec.whatwg.org/#concept-encoding-get\fP.
.sp
For more details on encoding and how ripgrep deals with it, see \fBGUIDE.md\fP.
.sp
The encoding detection that ripgrep uses can be reverted to its automatic mode
via the \fB\-\-no-encoding\fP flag.
.RE
.sp
\fB\-\-engine\fP=\fIENGINE\fP
.RS 4

Editor/Binaries/Linux/ripgrep-14.1.1-x86_64-unknown-linux-musl/complete/rg.bash
_rg() {
  local i cur prev opts cmds
  COMPREPLY=()
  cur="${COMP_WORDS[COMP_CWORD]}"
  prev="${COMP_WORDS[COMP_CWORD-1]}"
  cmd=""
  opts=""

  for i in ${COMP_WORDS[@]}; do
    case "${i}" in
      rg)
        cmd="rg"
        ;;
      *)
        ;;
    esac
  done

  case "${cmd}" in
    rg)
      opts="--regexp -e --file -f --after-context -A --before-context -B --binary --no-binary --block-buffered --no-block-buffered --byte-offset -b --no-byte-offset --case-sensitive -s --color --colors --column --no-column --context -C --context-separator --no-context-separator --count -c --count-matches --crlf --no-crlf --debug --dfa-size-limit --encoding -E --no-encoding --engine --field-context-separator --field-match-separator --files --files-with-matches -l --files-without-match --fixed-strings -F --no-fixed-strings --follow -L --no-follow --generate --glob -g --glob-case-insensitive --no-glob-case-insensitive --heading --no-heading --help -h --hidden -. --no-hidden --hostname-bin --hyperlink-format --iglob --ignore-case -i --ignore-file --ignore-file-case-insensitive --no-ignore-file-case-insensitive --include-zero --no-include-zero --invert-match -v --no-invert-match --json --no-json --line-buffered --no-line-buffered --line-number -n --no-line-number -N --line-regexp -x --max-columns -M --max-columns-preview --no-max-columns-preview --max-count -m --max-depth -d --max-filesize --mmap --no-mmap --multiline -U --no-multiline --multiline-dotall --no-multiline-dotall --no-config --no-ignore --ignore --no-ignore-dot --ignore-dot --no-ignore-exclude --ignore-exclude --no-ignore-files --ignore-files --no-ignore-global --ignore-global --no-ignore-messages --ignore-messages --no-ignore-parent --ignore-parent --no-ignore-vcs --ignore-vcs --no-messages --messages --no-require-git --require-git --no-unicode --unicode --null -0 --null-data --one-file-system --no-one-file-system --only-matching -o --path-separator --passthru --pcre2 -P --no-pcre2 --pcre2-version --pre --no-pre --pre-glob --pretty -p --quiet -q --regex-size-limit --replace -r --search-zip -z --no-search-zip --smart-case -S --sort --sortr --stats --no-stats --stop-on-nonmatch --text -a --no-text --threads -j --trace --trim --no-trim --type -t --type-not -T --type-add --type-clear --type-list --unrestricted -u --version -V --vimgrep --with-filename -H --no-filename -I --word-regexp -w --auto-hybrid-regex --no-auto-hybrid-regex --no-pcre2-unicode --pcre2-unicode --sort-files --no-sort-files <PATTERN> <PATH>..."
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
        COMPREPLY=($(compgen -W "${opts}" -- "${cur}"))
        return 0
      fi
      case "${prev}" in

        --regexp)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -e)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --file)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -f)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --after-context)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -A)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --before-context)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -B)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --binary)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-binary)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --block-buffered)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-block-buffered)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --byte-offset)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -b)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-byte-offset)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --case-sensitive)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -s)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --color)
          COMPREPLY=($(compgen -W "never auto always ansi" -- "${cur}"))
          return 0
          ;;
        --colors)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --column)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-column)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --context)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -C)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --context-separator)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-context-separator)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --count)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -c)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --count-matches)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --crlf)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-crlf)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --debug)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --dfa-size-limit)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --encoding)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -E)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-encoding)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --engine)
          COMPREPLY=($(compgen -W "default pcre2 auto" -- "${cur}"))
          return 0
          ;;
        --field-context-separator)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --field-match-separator)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --files)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --files-with-matches)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -l)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --files-without-match)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --fixed-strings)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -F)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-fixed-strings)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --follow)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -L)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-follow)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --generate)
          COMPREPLY=($(compgen -W "man complete-bash complete-zsh complete-fish complete-powershell" -- "${cur}"))
          return 0
          ;;
        --glob)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -g)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --glob-case-insensitive)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-glob-case-insensitive)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --heading)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-heading)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --help)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -h)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --hidden)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -.)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-hidden)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --hostname-bin)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --hyperlink-format)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --iglob)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-case)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -i)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-file)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-file-case-insensitive)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-file-case-insensitive)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --include-zero)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-include-zero)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --invert-match)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -v)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-invert-match)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --json)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-json)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --line-buffered)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-line-buffered)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --line-number)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -n)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-line-number)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -N)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --line-regexp)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -x)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --max-columns)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -M)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --max-columns-preview)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-max-columns-preview)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --max-count)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -m)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --max-depth)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -d)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --max-filesize)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --mmap)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-mmap)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --multiline)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -U)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-multiline)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --multiline-dotall)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-multiline-dotall)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-config)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-dot)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-dot)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-exclude)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-exclude)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-files)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-files)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-global)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-global)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-messages)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-messages)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-parent)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-parent)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-ignore-vcs)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --ignore-vcs)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-messages)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --messages)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-require-git)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --require-git)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-unicode)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --unicode)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --null)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -0)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --null-data)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --one-file-system)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-one-file-system)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --only-matching)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -o)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --path-separator)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --passthru)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --pcre2)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -P)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-pcre2)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --pcre2-version)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --pre)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-pre)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --pre-glob)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --pretty)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -p)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --quiet)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -q)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --regex-size-limit)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --replace)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -r)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --search-zip)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -z)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-search-zip)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --smart-case)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -S)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --sort)
          COMPREPLY=($(compgen -W "none path modified accessed created" -- "${cur}"))
          return 0
          ;;
        --sortr)
          COMPREPLY=($(compgen -W "none path modified accessed created" -- "${cur}"))
          return 0
          ;;
        --stats)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-stats)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --stop-on-nonmatch)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --text)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -a)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-text)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --threads)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -j)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --trace)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --trim)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-trim)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --type)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -t)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --type-not)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -T)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --type-add)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --type-clear)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --type-list)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --unrestricted)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -u)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --version)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -V)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --vimgrep)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --with-filename)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -H)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-filename)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -I)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --word-regexp)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -w)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --auto-hybrid-regex)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-auto-hybrid-regex)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-pcre2-unicode)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --pcre2-unicode)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --sort-files)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --no-sort-files)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
      esac
      COMPREPLY=($(compgen -W "${opts}" -- "${cur}"))
      return 0
      ;;
  esac
}

complete -F _rg -o bashdefault -o default rg
